ssvsigner/ekm/remote_key_manager.go

```diff
@@ -0,0 +1,463 @@
+		}
+
+	case spectypes.DomainSelectionProof:
+		data, ok := obj.(spectypes.SSZUint64)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to SSZUint64")
+		}
+
+		req.Type = web3signer.TypeAggregationSlot
+		req.AggregationSlot = &web3signer.AggregationSlot{Slot: phase0.Slot(data)}
+
+	case spectypes.DomainRandao:
+		data, ok := obj.(spectypes.SSZUint64)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to SSZUint64")
+		}
+
+		req.Type = web3signer.TypeRandaoReveal
+		req.RandaoReveal = &web3signer.RandaoReveal{Epoch: phase0.Epoch(data)}
+
+	case spectypes.DomainSyncCommittee:
+		data, ok := obj.(spectypes.SSZBytes)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to SSZBytes")
+		}
+
+		req.Type = web3signer.TypeSyncCommitteeMessage
+		req.SyncCommitteeMessage = &web3signer.SyncCommitteeMessage{
+			BeaconBlockRoot: phase0.Root(data),
+			Slot:            slot,
+		}
+
+	case spectypes.DomainSyncCommitteeSelectionProof:
+		data, ok := obj.(*altair.SyncAggregatorSelectionData)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to SyncAggregatorSelectionData")
+		}
+
+		req.Type = web3signer.TypeSyncCommitteeSelectionProof
+		req.SyncAggregatorSelectionData = &web3signer.SyncCommitteeAggregatorSelection{
+			Slot:              data.Slot,
+			SubcommitteeIndex: phase0.CommitteeIndex(data.SubcommitteeIndex),
+		}
+
+	case spectypes.DomainContributionAndProof:
+		data, ok := obj.(*altair.ContributionAndProof)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to ContributionAndProof")
+		}
+
+		req.Type = web3signer.TypeSyncCommitteeContributionAndProof
+		req.ContributionAndProof = data
+
+	case spectypes.DomainApplicationBuilder:
+		data, ok := obj.(*eth2apiv1.ValidatorRegistration)
+		if !ok {
+			return nil, phase0.Root{}, errors.New("could not cast obj to ValidatorRegistration")
+		}
+
+		req.Type = web3signer.TypeValidatorRegistration
+		req.ValidatorRegistration = data
+	default:
+		return nil, phase0.Root{}, errors.New("domain unknown")
+	}
+
+	root, err := spectypes.ComputeETHSigningRoot(obj, domain)
+	if err != nil {
+		return nil, phase0.Root{}, fmt.Errorf("compute root: %w", err)
+	}
+
+	req.SigningRoot = root
+
+	sig, err := km.signerClient.Sign(ctx, sharePubkey, req)
+	if err != nil {
+		return spectypes.Signature{}, phase0.Root{}, fmt.Errorf("remote signer: %w", err)
+	}
+
+	return sig[:], root, nil
+}
+
+func (km *RemoteKeyManager) handleDomainAttester(
+	obj ssz.HashRoot,
+	sharePubkey phase0.BLSPubKey,
+) (*phase0.AttestationData, error) {
+	data, ok := obj.(*phase0.AttestationData)
+	if !ok {
+		return nil, errors.New("could not cast obj to AttestationData")
+	}
+
+	if err := km.IsAttestationSlashable(sharePubkey, data); err != nil {
+		return nil, err
+	}
+
+	if err := km.UpdateHighestAttestation(sharePubkey, data); err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
```

Comment:
i think there's a risk for slashing here if there are parallel signature requests for the same pubkey

that's the reason eth2-key-manager locks pubkeys for the whole sequence of: `IsAttestationSlashable -> UpdateHighestAttestation -> Sign` (see https://github.com/ssvlabs/eth2-key-manager/blob/f6703e8353f6e64c377884acf651a30354a1075a/signer/sign_attestation.go#L22-L24)

and i think we should do the same here, unless there's a good reason not to

Reply:
copied, as well as far future checks